### DISCUSSION TITLE
Github API Collaborators access changed today

### DISCUSSION THREAD

**Discussion by omnibs** [community/community#24618] [Apps, API and Webhooks]
<p>It seems until yesterday accounts with <code>Write</code> role on repos could read <code>https://api.github.com/repos/[owner]/[repo]/collaborators</code> on the GH API.</p>
<p>Today, only the <code>Admin</code> role can.</p>
<p>Our <a href="https://github.com/rust-lang/homu" rel="noopener nofollow ugc">merge bot</a> relies on that for automatically figuring out who can trigger it.</p>
<p>Was this change a mistake on Github’s part? Is it going to be reverted?</p>
<p>If it’s intentional, is there some other recommended way for automation to detect collaborator permissions which doesn’t require the <code>Admin</code> role?</p>

**Comment by bertptrs** [comment]
<p>I found this issue while diagnosing a similar problem with our merge bot. This change started affecting us somewhere in the last two hours.</p>

**Comment by nethgato** [comment]
<p>Hey <a class="mention" href="/omnibs">@omnibs</a> and <a class="mention" href="/bertptrs">@bertptrs</a> :wave:</p>
<p>You all caught this well before I could! We did push a recent update (4 hours ago, as of this writing) that now requires <code>admin</code> access via the API to see collaborators.</p>
<p>This change brings our API in-line with the Web based experience, and should be more consistent.</p>
<p>There’s quite an extensive reasoning behind this and if it’s necessary I can elevate those details.</p>
<p>So yes, this was a deliberate action. For now, I don’t believe there is another permissions level outside of <code>admin</code> to perform this action.</p>
<p>Alternatively, it is true that GitHub Integrations have the ability to list collaborators, but we will likely be reviewing that in the future.</p>
<p>Edit: our team is currently in discussion about the impact this might have, and whether or not a reversion is warranted. Will update when there’s more to share.</p>

**Comment by pedrobaeza** [comment]
<p>This change is affecting our merge bot that requires such permission for listing the authorized people to perform such action:</p>
<aside class="onebox githubpullrequest" data-onebox-src="https://github.com/OCA/OpenUpgrade/pull/2854#issuecomment-920041953">
  <header class="source">

      <a href="https://github.com/OCA/OpenUpgrade/pull/2854#issuecomment-920041953" target="_blank" rel="noopener nofollow ugc">github.com/OCA/OpenUpgrade</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Pull Request">
    <svg width="60" height="60" class="github-icon" viewbox="0 0 12 16" aria-hidden="true"><path d="M11 11.28V5c-.03-.78-.34-1.47-.94-2.06C9.46 2.35 8.78 2.03 8 2H7V0L4 3l3 3V4h1c.27.02.48.11.69.31.21.2.3.42.31.69v6.28A1.993 1.993 0 0 0 10 15a1.993 1.993 0 0 0 1-3.72zm-1 2.92c-.66 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2zM4 3c0-1.11-.89-2-2-2a1.993 1.993 0 0 0-1 3.72v6.56A1.993 1.993 0 0 0 2 15a1.993 1.993 0 0 0 1-3.72V4.72c.59-.34 1-.98 1-1.72zm-.8 10c0 .66-.55 1.2-1.2 1.2-.65 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2zM2 4.2C1.34 4.2.8 3.65.8 3c0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/OCA/OpenUpgrade/pull/2854" target="_blank" rel="noopener nofollow ugc">[14.0][MIG] im_livechat</a>
    </h4>

    <div class="branches">
      <code>OCA:14.0</code> ← <code>ForgeFlow:14.0-mig-im_livechat</code>
    </div>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2021-07-22" data-time="08:51:38" data-timezone="UTC">08:51AM - 22 Jul 21 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/MiquelRForgeFlow" target="_blank" rel="noopener nofollow ugc">
          <img alt="MiquelRForgeFlow" src="https://user-images.githubusercontent.com/7165771/181085674-f88b23d3-dfae-46e1-961d-8d3ae1136d3a.jpeg" class="onebox-avatar-inline" width="20" height="20">
          MiquelRForgeFlow
        </a>
      </div>

      <div class="lines" title="1 commits changed 4 files with 43 additions and 1 deletions">
        <a href="https://github.com/OCA/OpenUpgrade/pull/2854/files" target="_blank" rel="noopener nofollow ugc">
          <span class="added">+43</span>
          <span class="removed">-1</span>
        </a>
      </div>
    </div>
  </div>
</div>

  <div class="github-row">
    <p class="github-body-container">Migration done.</p>
  </div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>If such list is public, why it should be restricted to admins?</p>

**Comment by emin63** [comment]
<p>Was there some warning or notification before this went live? Is there a way to be warned about similar changes which could break tools using the GitHub API in the future?</p>
<p>Thanks</p>

**Comment by nethgato** [comment]
<p>Hi <a class="mention" href="/emin63">@emin63</a> :wave:</p>
<p>Thanks for joining the Support Community to participate in this thread.</p>
<p>Unfortunately, this change was made silently, after comparing the experience with accessing this information via the web UI, and the API. As it was, it was an inconsistent experience where only Administrators could view collaborator information on the UI, but <code>push</code> access could grant these privileges for the API.</p>
<p>In my edit earlier I mentioned we were considering a rollback, but that doesn’t seem likely now.</p>
<p>We should be posting to our blog about this soon. Though I can’t say exactly when.</p>

**Comment by matteoopenf** [comment]
<p>I have the same problem</p>

**Comment by hatboysam** [comment]
<p>I believe this may have also broken the “Get Repository Permissions for a User” API even when a user is trying to access their OWN permissions:<br>
<a href="https://docs.github.com/en/rest/reference/repos#get-repository-permissions-for-a-user" class="onebox" target="_blank" rel="noopener nofollow ugc">https://docs.github.com/en/rest/reference/repos#get-repository-permissions-for-a-user</a></p>
<p>I don’t think a user should need to be admin to check their own permissions. Is there any other way for a user to use the API to find out their permission level on a repo?</p>

**Comment by nethgato** [comment]
<p>Hey <a class="mention" href="/hatboysam">@hatboysam</a> :wave:</p>
<p>I’ve followed the discussion since this morning, and that portion of our codebase (get permission) should not have been impacted. At least, that’s my current understanding.</p>
<p>Would you mind submitting this as <a href="https://docs.github.com/en/github/working-with-github-support/submitting-a-ticket">a Support ticket</a>? Or if you prefer, we can split this off into it’s own thread.</p>
<p>For now, increasing permissions for these automation tools to <code>admin</code> is going to be the solution.</p>
<p>I’m so sorry for the lack of comms before this was put out, ya’ll. Not the best experience here.</p>

**Comment by hatboysam** [comment]
<p>Thanks for the quick response. I filed ticket 1310760 a few hours ago, it has some details. I was able to work around this by using the “get repository” API which returns a personalized <code>permissions</code> field since in my application users only need to know their own permissions.</p>
<p>But it would be great if you could investigate anyway, as I do believe that API is now behaving as I described.</p>

**Comment by nethgato** [comment]
<p>Went back and yes, the <code>collaborators/{user}/permission</code> endpoint for <code>/repos</code>, has also been updated to only allow the <code>owner</code> access to permission details. So if you are a <code>collaborator</code>, you can no longer view this via the API.</p>
<p>Again, apologies for the lack of transparency before this change was made. I myself wasn’t aware until this thread.</p>
<p>We hope to have out a /changelog post soon. :crossed_fingers:</p>

**Comment by FPtje** [comment]
<aside class="quote group-github-employees" data-username="nethgato" data-post="9" data-topic="201259">
<div class="title">
<div class="quote-controls"></div>
 nethgato:</div>
<blockquote>
<p>For now, increasing permissions for these automation tools to <code>admin</code> is going to be the solution.</p>
</blockquote>
</aside>
<p>Requiring admin privilege merely for checking whether a user would be allowed to merge a pull request seems overbearing. This would go directly against the principle of least privilege. This change would mean that bots will have to be guarded against accidentally doing any of the other catastrophic actions that admin permissions would bring.</p>
<p>Bots need a non-admin way to know whether a user can trigger it. The jerk change to suddenly make merge bots admin is going to make them a prime target for attack. No one saw this coming, and suddenly giving merge bots admin access is going to be a huge and very obvious security risk.</p>
<p>Not communicating this change isn’t the biggest problem, the biggest problem here is demanding merge bot users to accept this security liability.</p>

**Comment by thaJeztah** [comment]
<aside class="quote no-group" data-username="FPtje" data-post="12" data-topic="201259">
<div class="title">
<div class="quote-controls"></div>
 FPtje:</div>
<blockquote>
<p>Requiring admin privilege merely for checking whether a user would be allowed to merge a pull request seems overbearing. This would go directly against the principle of least privilege. This change would mean that bots will have to be guarded against accidentally doing any of the other catastrophic actions that admin permissions would bring.</p>
</blockquote>
</aside>
<p>:point_up: this; we also got hit by this change, effectively breaking all our Jenkins jobs.</p>

**Comment by fredg02** [comment]
<aside class="quote no-group" data-username="FPtje" data-post="12" data-topic="201259">
<div class="title">
<div class="quote-controls"></div>
 FPtje:</div>
<blockquote>
<p>Requiring admin privilege merely for checking whether a user would be allowed to merge a pull request seems overbearing. This would go directly against the principle of least privilege. This change would mean that bots will have to be guarded against accidentally doing any of the other catastrophic actions that admin permissions would bring.</p>
</blockquote>
</aside>
<p>Fully agree. Several Eclipse projects are affected by this as well. See also <a href="https://www.eclipse.org/lists/cross-project-issues-dev/msg18641.html" class="inline-onebox" rel="noopener nofollow ugc">[cross-project-issues-dev] GitHub issue - "Must have push access to view</a></p>

**Comment by mkumatag** [comment]
<p>Even it is effecting us running <a href="https://github.com/kubernetes/test-infra/tree/master/prow" rel="noopener nofollow ugc">prow</a> bot</p>

**Comment by hatboysam** [comment]
<p><a class="mention" href="/nethgato">@nethgato</a> you said this change was made to bring the API in line with the information accessible via the Web UI, but actually it is not.</p>
<p>My use case for this API is to allow users to find someone who can review their code. On GitHub this is done via the Reviewers autocomplete box. This box only allows the user to pick <strong>other collaborators</strong> and is visible to anyone with write acccess:</p>
<p><img src="https://user-images.githubusercontent.com/997280/181085681-33b9c2db-5150-4a9a-8844-68bf708b0cd8.png" alt="image" data-base62-sha1="5yFcttz7gCxSSycVXDACzlKgwjc" width="327" height="199"></p>
<p>So this information is already intentionally visible in the Web UI to anyone with write access. I would really appreciate if the API would give me at least the same ability to list collaborators.</p>

**Comment by nethgato** [comment, marked as answer]
<p>Hi all!</p>
<p>While I was away, a larger number of users both wrote into this thread and into our Support ticketing system. After taking into the consideration the experience this was having, our engineering team rolled back this update.</p>
<p>I’m so sorry for the confusing and frustrating impact this had.</p>
<p>When this comes back up to review for re-adding to our API, we will be posting to the changelog in advance of the change.</p>

**Comment by pedrobaeza** [comment]
<p>Thanks for reconsidering.</p>

**Comment by matteoopenf** [comment]
<p>Many Thanks.<br>
Nice solution</p>

**Comment by nethgato** [comment]
<p>Ah, I shouldn’t take any thanks here, but I appreciate how perplexing it must have been to see these 403s out of nowhere and without a changelog to point to.</p>
<p>And I should further clarify (rather than edit my last post) that I should not say “when this comes back up,” but “if.”</p>
<p>We are considering marking this internally as a “won’t fix,” but there are some internal conversations that initially sparked the change, and was fast tracked, without considering the impact on automation tools that leverage those endpoints.</p>
<p>So <em>if</em> it comes back, it will be announced in advance.</p>

**Comment by Jessa1233** [comment]
Hey [@omnibs](https://github.com/omnibs) and [@bertptrs](https://github.com/bertptrs) 👋

You all caught this well before I could! We did push a recent update (4 hours ago, as of this writing) that now requires admin access via the API to see collaborators.

This change brings our API in-line with the Web based experience, and should be m[o](https://xxxtentacionmerchshop.net/)re consistent.

There’s quite a[n](https://bapehoodieofficial.net/) extensive reasoning b[e](https://trapstarclothingshop.com/bags/)hind this and if it’s necessary I c[a](https://weekndmerchshop.com/)n elevate those details.

So yes, this was a deliberate action. For n[o](https://kanyewestmerchshop.ltd/)w, I don’t believe there is a[n](https://playboycartimerch.com/)other permi[s](https://trapstarclothingshop.com/tracksuit/)sions le[v](https://dreammerchshop.com/)el outside of admin to perform this action.

Alternatively, it is true that GitHub Integrations have the ability to list collaborat[o](https://trapstarclothingshop.com/t-shirts/)rs, but we will likely be reviewing that in the future.

Edit: o[u](https://trapstarclothingshop.com/hoodies/)r team is currently in discussion about the im[p](https://trapstarclothingshop.com/shorts/)act this might have, and whether [o](https://trapstarclothingshop.com/)r not a reversion is warranted. Will update when there’s more to share.
